src/vs/editor/contrib/documentSymbols/outlineModel.ts

```diff
@@ -88,18 +90,18 @@ export abstract class TreeElement {
 export class OutlineElement extends TreeElement {
 
 	children: { [id: string]: OutlineElement; } = Object.create(null);
-	score: FuzzyScore = FuzzyScore.Default;
-	marker: { count: number, topSev: MarkerSeverity };
+	score: FuzzyScore | undefined = FuzzyScore.Default;
+	marker: { count: number, topSev: MarkerSeverity } | undefined;
 
 	constructor(
 		readonly id: string,
-		public parent: OutlineModel | OutlineGroup | OutlineElement,
+		public parent: TreeElement | undefined,
```

Comment:
please refrain from making changes aren't needed for strict null checks

Reply:
I was surprised and spent quite a while trying to understand why, but this use of inheritance did cause errors under strict null checks. You can see them by checking out commit `9bbe60f` (I separated the 'normal' null error fixes from the inheritance-related ones in separate commits).

```
src/vs/editor/contrib/documentSymbols/outlineModel.ts:90:2 - error TS2416: Property 'children' in type 'OutlineElement' is not assignable to the same property in base type 'TreeElement'.
  Type '{ [id: string]: OutlineElement; }' is not assignable to type '{ [id: string]: TreeElement; }'.
    Index signatures are incompatible.
      Type 'OutlineElement' is not assignable to type 'TreeElement'.
        Types of property 'parent' are incompatible.
          Type 'OutlineElement | OutlineModel | OutlineGroup' is not assignable to type 'TreeElement'.
            Type 'OutlineModel' is not assignable to type 'TreeElement'.
              Types of property 'children' are incompatible.
                Type '{ [id: string]: OutlineElement | OutlineGroup; }' is not assignable to type '{ [id: string]: TreeElement; }'.
                  Index signatures are incompatible.
                    Type 'OutlineElement | OutlineGroup' is not assignable to type 'TreeElement'.
                      Type 'OutlineGroup' is not assignable to type 'TreeElement'.
                        Types of property 'adopt' are incompatible.
                          Type '(parent: OutlineModel) => OutlineGroup' is not assignable to type '(newParent: TreeElement) => TreeElement'.
                            Types of parameters 'parent' and 'newParent' are incompatible.
                              Type 'TreeElement' is missing the following properties from type 'OutlineModel': _groups, textModel, _compact, merge, and 5 more.

90  children: { [id: string]: OutlineElement; } = Object.create(null);
```

I'm not 100% confident in my understand of this error but I think the error prevents a hypothetical error that could be make like this:
```
class SomeNewTypeOfTreeElement {
  ...
}

const tree: TreeElement = new OutlineElement(...)
tree.parent = new SomeNewTypeOfTreeElement(...) // this works because TreeElement doesn't know about OutlineElement
tree.doSomethingWithParentThatOutlineElementDoesNotExpect()
```

There's no such code right now, of course, but the type system doesn't care. Happy to learn if there's a better way to silence these errors though.

Reply:
Hm, intersting. Again, in doubt try to use the `!` operator. My challenge is that it's hard to reason about these things but that I know that the code in its current shape runs. So, I'd prefer to not really change it

Reply:
I understand, but this just changes the type annotation from `OutlineModel | OutlineGroup | OutlineElement` to their parent `TreeElement`, so I think it's safe? See https://github.com/Microsoft/vscode/pull/65062/commits/38d4a547d7f219e83d5e8a922f5ba85f283d7a86